src/noit_filters.c

```diff
@@ -655,6 +685,76 @@ noit_apply_filterrule_metric(filterrule_t *r,
   return mtev_true;
 }
 
+static int
+noit_add_measurement_tag(filterrule_t *r,
+                         metric_t *metric,
+                         char **expanded_metric_name,
+                         noit_metric_tagset_t *mtset) {
+  if (!r || !metric || !expanded_metric_name || !mtset || !r->measurement_tag.add_measurement_tag_cat || mtset->tag_count >= MAX_TAGS) {
+    return -1;
+  }
+  char encoded_nametag[NOIT_TAG_MAX_PAIR_LEN+1];
+  char decoded_nametag[NOIT_TAG_MAX_PAIR_LEN+1];
+  const char *cat = r->measurement_tag.add_measurement_tag_cat;
+  const char *val = r->measurement_tag.add_measurement_tag_val ? r->measurement_tag.add_measurement_tag_val : "";
+  snprintf(decoded_nametag, sizeof(decoded_nametag), "%s%c%s", cat, NOIT_TAG_DECODED_SEPARATOR, val);
+  size_t nlen = noit_metric_tagset_encode_tag(encoded_nametag, sizeof(encoded_nametag),
+                                              decoded_nametag, strlen(decoded_nametag));
+  for (int i = 0; i < mtset->tag_count; i++) {
+    if (nlen == mtset->tags[i].total_size && !strncmp(encoded_nametag, mtset->tags[i].tag, nlen)) {
+      /* tag is already there */
+      return 1;
+    }
+  }
+
+  /* Extend the metric name */
+  if (!(*expanded_metric_name)) {
+    *expanded_metric_name = strdup(metric->metric_name);
+  }
+  size_t old_size = strlen(*expanded_metric_name);
+  size_t encoded_nametag_size = strlen(encoded_nametag);
+  /* `+5` is for |MT{} */
+  size_t new_size = old_size + encoded_nametag_size + 5 + 1;
+  *expanded_metric_name = realloc(*expanded_metric_name, new_size);
+  mtevAssert(*expanded_metric_name);
+  memcpy(*expanded_metric_name + old_size, "|MT{", 4);
+  memcpy(*expanded_metric_name + old_size + 4, encoded_nametag, encoded_nametag_size);
+  (*expanded_metric_name)[new_size - 2] = '}';
+  (*expanded_metric_name)[new_size - 1] = 0;
+
+  mtset->tags[mtset->tag_count].category_size = strlen(r->measurement_tag.add_measurement_tag_cat + 1);
+  mtset->tags[mtset->tag_count].total_size = nlen;
+  mtset->tags[mtset->tag_count].tag = strdup(encoded_nametag);
+  mtset->tag_count++;
+
+  return 0;
+}
+
+static void
+noit_update_metric_name(char *expanded_metric_name,
+                        noit_metric_tagset_t *mtset,
+                        int mtset_tag_start,
+                        metric_t *metric) {
+  char buff[MAX_METRIC_TAGGED_NAME];
+  char *old = metric->metric_name;
+  if (!expanded_metric_name) {
+    expanded_metric_name = strdup(old);
+  }
+  if (noit_metric_canonicalize(expanded_metric_name, strlen(expanded_metric_name), buff, sizeof(buff), mtev_true)) {
```

Comment:
Can you check ``noit_metric_canonicalize_ex`` while you're here?  I think it can overflow its output argument by 1 byte.


```c
  if(null_term) output[out-buff] = '\0';
  if((out-buff) > MAX_METRIC_TAGGED_NAME) return -8;
```